Add logic to BackgroundProcessor to allow dynamically switching modes elegantly, and deprecate pre-existing interfaces

1egoman opened the pull request:
In a [past pull request](https://github.com/livekit/track-processors-js/pull/95), I had attempted to come up with a more elegant way to dynamically switch between track processor modes, but got the feedback what I was proposing was a bit too large of a leap and not backwards compatible enough.

### Summary

So, I've attempted to do this again in this change, only via a bit of different mechanism:
1. I've updated `BackgroundProcessor` to now optionally take a `mode` key when being initially set up to define an initial starting mode. For example: `BackgroundProcessor({ mode: 'background-blur', blurRadius: BLUR_RADIUS })` or `BackgroundProcessor({ mode: 'virtual-background', imagePath: "..." })`.
2. I've also updated `BackgroundProcessor`'s return value - it now returns a new superclass of `ProcessorWrapper` called `BackgroundProcessorWrapper` which exposes `switchToBackgroundBlur` and `switchToVirtualBackground` functions, which allows for more convenient switching as opposed to calling `updateTransformerOptions` manually.

In addition, per some [other](https://github.com/livekit/track-processors-js/issues/105) [discussions](https://github.com/livekit/track-processors-js/issues/85) I've had in other issues / pull requests, I wanted to put a demo together that would showcase an experience which wouldn't have any flicker due to media pipeline delay (technical info about this can be found [here](https://github.com/livekit/track-processors-js/pull/96)). In order to accomplish this, I've added a "disabled" mode to the `BackgroundProcessor` - so there now is a new `BackgroundProcessor({ mode: 'disabled' })` constructor signature, and also a new `switchToDisabled` method on the `BackgroundProcessorWrapper`.

This means that a user can effectively get rid of all "flicker" by enable the background processor immediately after connecting in the `disabled` mode, and then later on since the media pipeline is all set up, there will be no artifacts when enabling either effect.

### Demo
I also updated the demo - previously, the "toggle buttons" provided for a quite weird interface for both controlling whether the processor was enabled / disabled, and also setting which mode it is in. Now, these are controlled seperately.


https://github.com/user-attachments/assets/b9fe812c-dab8-4367-81e9-151ed9a30ca8

xianshijing-lk commented on example/sample.ts:
```diff
@@ -246,43 +242,57 @@ const appActions = {
     videoPub.videoTrack?.restartTrack(options);
   },
 
+  toggleTrackProcessorEnabled: async () => {
```
It looks like ~80% of code in the following 4 functions are duplicated.
toggleTrackProcessorEnabled
toggleBlur
toggleDisabledBackground
toggleVirtualBackground

can we try consolidate them into 1 or 2 helper function to make things easier ? not sure if things like 
switchFn: (bg: any) => Promise<void>
is acceptable from livekit JS style guide.

xianshijing-lk commented on example/sample.ts:
```diff
@@ -748,6 +783,42 @@ function updateButtonsForPublishState() {
+    'virtual-background': { active: 'toggle-virtual-bg-button', inactive: ['toggle-disabled-button', 'toggle-blur-button'] },
+    'background-blur': { active: 'toggle-blur-button', inactive: ['toggle-virtual-bg-button', 'toggle-disabled-button'] },
+    'off': { active: null, inactive: [] },
+  }[state.backgroundMode ?? 'off'];
```
quick question, will backgroundMode be something else ? like not disabled, virtual-background, background-blur, off, nil, undefined ? 

If that is possible, I will it will throw. Maybe that is expected behavior, I just want to make sure.

xianshijing-lk commented on src/index.ts:
```diff
@@ -27,16 +29,179 @@ export const supportsBackgroundProcessors = () =>
+
+  // Extract transformer-specific options and processor options
+  let transformer, processorOpts;
+  switch (options.mode) {
```
will it work if creating a small helper to extract the transformer options so that we can remove duplicated code ? 

const transformerOptions = (() => {
  switch (options.mode) {
    case 'background-blur':
      return { blurRadius: options.blurRadius, segmenterOptions: options.segmenterOptions, assetPaths: options.assetPaths, onFrameProcessed: options.onFrameProcessed };
    case 'virtual-background':
      return { imagePath: options.imagePath, segmenterOptions: options.segmenterOptions, assetPaths: options.assetPaths, onFrameProcessed: options.onFrameProcessed };
    case 'disabled':
      return { segmenterOptions: options.segmenterOptions, assetPaths: options.assetPaths, onFrameProcessed: options.onFrameProcessed };
    default:
      return { blurRadius: options.blurRadius, imagePath: options.imagePath, segmenterOptions: options.segmenterOptions, assetPaths: options.assetPaths, onFrameProcessed: options.onFrameProcessed };
  }
})();
processorOpts = { ...options };
transformer = new BackgroundTransformer(transformerOptions);

lukasIO commented on example/sample.ts:
```diff
@@ -34,12 +34,8 @@ const IMAGE_PATH = '/samantha-gades-BlIhVfXbi9s-unsplash.jpg';
-  virtualBackground: VirtualBackground(IMAGE_PATH, undefined, (stats) => {
-    // console.log('frame processing stats', stats);
-  }),
+  backgroundMode: null as NonNullable<BackgroundProcessorOptions['mode']> | null,
```
Instead of tracking this within the sample, how about holding the current mode (and exposing a way to retrieve it) on the `BackgroundProcessorWrapper` directly?

1egoman replied:
I thought about doing this, but the reason I didn't was because if I did the bookkeeping in `BackgroundProcessorWrapper`, then somebody could just call `updateTransformerOptions` themselves and get it into a broken state.

The concern you are bringing up is valid though and I think it would make the logic much cleaner - let me think about how to best accomplish this. Maybe I could override `updateTransformerOptions` and do that check at that level to guard against this 🤔

1egoman replied:
I thought about it and added a `mode` getter which inspects the options and computes the mode on the fly. I think that's the least intrusive option!

lukasIO commented on example/sample.ts:
```diff
@@ -34,12 +34,8 @@ const IMAGE_PATH = '/samantha-gades-BlIhVfXbi9s-unsplash.jpg';
-    // console.log('frame processing stats', stats);
-  }),
+  backgroundMode: null as NonNullable<BackgroundProcessorOptions['mode']> | null,
+  background: BackgroundProcessor({ mode: 'background-blur', blurRadius: BLUR_RADIUS }),
```
Nit
```suggestion
  backgroundProcessor: BackgroundProcessor({ mode: 'background-blur', blurRadius: BLUR_RADIUS }),
```

1egoman replied:
💯 Not sure how I let that slip through, thanks!

lukasIO commented on src/index.ts:
```diff
@@ -27,16 +29,179 @@ export const supportsBackgroundProcessors = () =>
+| BackgroundProcessorLegacyOptions;
+
+class BackgroundProcessorWrapper extends ProcessorWrapper<BackgroundOptions> {
+  async switchToBackgroundBlur(blurRadius: number = DEFAULT_BLUR_RADIUS) {
```
Have you considered a `switchTo(mode, ...opts?)` method? 

Not saying that would necessarily be preferable, just curious about your thoughts on one vs the other.

1egoman replied:
I don't think I have a strong preference. A few unstructured thoughts:
- In an IDE of some sort, if you type `backgroundProcessor.` and let the autocomplete pop up, in the current state you'd see all the options, versus just `switchTo`
- At least initially, `switchTo` seems like it would be nice because you could parameterize the values being passed in (ie, instead of all the big `switch (state.backgroundMode)` type stuff in the example app or similar logic in a user's app, you could somehow just flatten that down to one `switchTo` call), but in practice this doesn't really work out because the overloads would make doing that in a type safe way challenging.
  - I think the better way to do `switchTo` might be `switchTo(opts: BackgroundProcessorOptions)` - so that would look like `backgroundProcessor.switchTo({ mode: 'background-blur', blurRadius: 10 })`.

Typing all that out, there's maybe an argument to have both? Especially since one is more discoverable and one is more flexible. I think I'll add both for now and see if anybody strongly opposes doing that.

1egoman commented:
> The sample is currently in a bit of an inbetween state with both the new insert and switch buttons visible but also the toggleBackground still being available as a separate button (albeit always disabled)

Ah that's what that button was for, I hadn't realized it was old. Removed.

> and the update background image button also being in a separate section of the UI.

Fair enough, moved it into the group!

> I don't know if this particular PR introduced this, but there's currently a flash of the "real" background if you click flipCam > which restarts the track.
> I think that's rather unexpected.

That is not something this pull request introduced. That being said, on cursory examination, it's for a very similar reason as this: https://github.com/livekit/track-processors-js/pull/96. I bet something similar to that fix could be adapted into the client sdk as part of `LocalTrack.restart` to make that at least somewhat smoother... 🤔
